lib/ukrand/random.c

```diff
@@ -0,0 +1,82 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/random.h>
+#include <string.h>
+
+static int uk_hwrand_init(struct uk_init_ctx *ictx __unused)
+{
+	return ukarch_random_init();
+}
+uk_early_initcall(uk_hwrand_init, 0x0);
+
+#if CONFIG_LIBUKRAND_HW
+int uk_random_u32(__u32 *val)
+{
+	return ukarch_random_u32(val);
+}
+
+int uk_random_u64(__u64 *val)
+{
+	return ukarch_random_u64(val);
+}
+
+int uk_random_seed_u32(__u32 *val)
+{
+	return ukarch_random_seed_u32(val);
+}
+
+int uk_random_seed_u64(__u64 *val)
+{
+	return ukarch_random_seed_u64(val);
+}
+#endif /* CONFIG_LIBUKRAND_HW */
+
+#if CONFIG_LIBUKRAND_CHACHA
+int uk_random_u32(__u32 *val)
+{
+	*val = uk_chacha_randr_u32();
+	return 0;
+}
+
+int uk_random_u64(__u64 *val)
+{
+	*val = uk_chacha_randr_u64();
+	return 0;
+}
+
+int uk_random_seed_u32(__u32 *val)
+{
+	*val = uk_chacha_randr_u32();
+	return 0;
+}
+
+int uk_random_seed_u64(__u64 *val)
+{
+	*val = uk_chacha_randr_u64();
+	return 0;
+}
+#endif /* CONFIG_LIBUKRAND_CHACHA */
+
+__ssz uk_random_fill_buffer(void *buf, __sz buflen)
+{
+	__sz step, chunk_size, i;
+	__u64 rd;
+
+	step = sizeof(__u64);
+	chunk_size = buflen % step;
+
+	for (i = 0; i < buflen - chunk_size; i += step)
+		uk_random_u64((__u64 *)((char *)buf + i));
```

Comment:
>  They seem to return 0 if the data in the destination is valid and 1 otherwise.

Based on comments from `uk/arch/random.h` they return a negative value. I will return that value if anything fails.

Reply:
Well then I believe the comments might be wrong. Even the return value despite being `int`, it is actually returned from an unsigned value and the instruction itself returning that value gives a positive result. Or am I missing something?

Reply:
I think you are right, the comments are wrong. I will add a commit soon to update the return value for `ukarch_random_*`,

Reply:
Done. I actually added it as two separate commits to better highlight changes, but we can always squash them.